src/main/scala/org/broadinstitute/dsde/workbench/sam/dataAccess/PostgresAccessPolicyDAO.scala

```diff
@@ -932,75 +980,103 @@ class PostgresAccessPolicyDAO(protected val writeDbRef: DbReference, protected v
 
   // Abstracts logic to load and unmarshal one or more policies, use to get full AccessPolicy objects from Postgres
   private def listPolicies(resource: FullyQualifiedResourceId, limitOnePolicy: Option[AccessPolicyName] = None, samRequestContext: SamRequestContext): IO[LazyList[AccessPolicy]] = {
+    readOnlyTransaction("listPolicies", samRequestContext)({ implicit session =>
+      // query to get policies and all associated roles and actions
+      val policyInfos = policyInfoQuery(resource, limitOnePolicy)
+
+      policyInfos.map { case (policyInfo, resultsByPolicy) =>
+        val (policyRoles, policyActions, policyDescendantPermissions) = unmarshalPolicyPermissions(resultsByPolicy.map(_._2))
+
+        val policyId = FullyQualifiedPolicyId(resource, policyInfo.name)
+        // queries to get all members as WorkbenchSubjects, there are 3 kinds, groups, users and policies
+        val memberGroups = directMemberGroupEmailsQuery(policyId).list().apply().map(_.name).toSet[WorkbenchSubject]
+        val memberUsers = directMemberUserEmailsQuery(policyId).list().apply().map(_.id).toSet[WorkbenchSubject]
+        val memberPolicies = directMemberPolicyIdAndEmailsQuery(policyId).list().apply().map(p => FullyQualifiedPolicyId(FullyQualifiedResourceId(p.resourceTypeName, p.resourceId), p.policyName)).toSet[WorkbenchSubject]
```

Comment:
The gist of this change is the highlighted code. The new code does a query for policy info that does not include membership info then does a follow on query for each of the 3 types of members. This means 1 query to get all the policies then 3 queries per policy. Now that I think about it though, if we like this, this could easily change to just 3 additional queries instead of 3n.

Contrast this to the previous state that did all this in 1 query with 5 more left joins and complicated handling of results. Complicated enough that I, who knows what the query is doing, was nervous touching it or adding to it to get the new information required by the use case.

From a performance perspective, the new code is probably worse but probably on the order of 10 to 20 ms (traces show the existing query takes ~3 ms loading a single policy). But I think we gain more maintainable code. And I think this approach is more stable as in I can think of some pathological examples that would cause some cartesian product discomfort in the existing query.

The other benefit is that the same queries to get policy info in one way are reused to get policy info in the form that the use case requires.

Reply:
also note that the existing queries was often followed by queries to load member email addresses which is not required in the proposed code